gui/biomes.lua

```diff
@@ -0,0 +1,392 @@
+        })
+    end
+    self.subviews.list:setChoices(choices)
+end
+
+do -- implementation of onMouseHoverEntry(idx, option)
+    function BiomeVisualizerLegend:onRenderFrame(dc, rect)
+        BiomeVisualizerLegend.super.onRenderFrame(self, dc, rect)
+
+        local list = self.subviews.list
+        local currentHoverIx = list:getIdxUnderMouse()
+        local oldIx = self.HoverIndex
+        if currentHoverIx ~= oldIx then
+            self.HoverIndex = currentHoverIx
+            if self.onMouseHoverEntry then
+                local choices = list:getChoices()
+                self:onMouseHoverEntry(currentHoverIx, choices[currentHoverIx])
+            end
+        end
+    end
+end
+
+local function add_field_text(lines, biome, field_name)
+    lines[#lines+1] = ("%s: %s"):format(field_name, biome[field_name])
+    lines[#lines+1] = NEWLINE
+end
+
+function BiomeVisualizerLegend:onMouseHoverEntry(idx, option)
+    if not idx then
+        self:ShowTooltip(nil)
+        return
+    end
+
+    local text = {}
+    text[#text+1] = ("type: %s"):format(df.biome_type[option.biomeTypeId])
+    text[#text+1] = NEWLINE
+
+    local biome = option.biome
+
+    add_field_text(text, biome, "savagery")
+    add_field_text(text, biome, "evilness")
+    table.insert(text, NEWLINE)
+
+    add_field_text(text, biome, "elevation")
+    add_field_text(text, biome, "rainfall")
+    add_field_text(text, biome, "drainage")
+    add_field_text(text, biome, "vegetation")
+    add_field_text(text, biome, "temperature")
+    add_field_text(text, biome, "volcanism")
+    table.insert(text, NEWLINE)
+
+    local flags = biome.flags
+    if flags.is_lake then
+        text[#text+1] = "lake"
+        text[#text+1] = NEWLINE
+    end
+    if flags.is_brook then
+        text[#text+1] = "brook"
+        text[#text+1] = NEWLINE
+    end
+
+    self:ShowTooltip(text)
+end
+
+function BiomeVisualizerLegend:ShowTooltip(text)
+    self.TooltipWindow = self.TooltipWindow or self.parent_view.subviews.legend_tooltip
+    self.TooltipWindow:ShowTooltip(text)
+end
+
+--------------------------------------------------------------------------------
+
+if RELOAD then TooltipWindow = nil end
+TooltipWindow = defclass(TooltipWindow, widgets.Window)
+
+local function calc_tooltip_frame(frame)
+    local frame = copyall(frame)
+    frame.t = frame.t + frame.h
+    frame.h = 15
+    return frame
+end
+
+TooltipWindow.ATTRS {
+    visible = false,
+    frame_title=DEFAULT_NIL,
+    frame_inset=0,
+    resizable=false,
+    frame_style = gui.PANEL_FRAME,
+    autoarrange_subviews = true,
+    autoarrange_gap = 0,
+    frame = calc_tooltip_frame(BiomeVisualizerLegend.ATTRS.frame),
+    --
+    parent_window = DEFAULT_NIL,
+}
+
+function TooltipWindow:init()
+    self:addviews{
+        widgets.Label{
+            view_id = 'label',
+            frame = {t = 0, h = 1000},
+            auto_height = false,
+            --wtf??? without this the label is always a single line
```

Comment:
this means that `updateLayout()` calls are not getting to this widget. If you call the `TooltipWindow`'s `updateLayout()` when the main panel's `updateLayout()` is called, then they should stay in sync. However, resizing from the line between the two "windows" is not a common UX style, and nothing else in DFHack does it that way. Maybe you could add the tooltip as a widget in `BiomeVisualizerLegend`, like how `gui/control-panel` does it? That would simplify the structure quite a bit. You can still specify an interior frame (`gui.INTERIOR_FRAME`) around the tooltip panel, if you wish.

Reply:
I was trying to recreate how hotkeys look, i.e.
![image](https://github.com/DFHack/scripts/assets/348150/9c88183f-4c90-4f31-91dd-3420fe6bd341)
but also hiding the tooltip if not hovering. I'll give it another try :)

Reply:
Note that the hotkeys panel is not resizable -- nor is the buildingplan panel, which also has that dual-frame look. For a resizable window, I think what you want is a single exterior frame with possibly a thinner interior frame around the tooltip. Otherwise it isn't obvious where the player should go to resize the window. You can still make the extended info disappear and the window shrink when it's not hovered, but I'd actually recommend leaving it. The player can always resize the window or dragging it around if it's in the way.

Reply:
I can't figure out how to make the window shrink dynamically :(
I'll try a version just showing the info

Reply:
I give up on the resizable version.
List is fix 5 rows, info is fix 15 rows.
![image](https://github.com/DFHack/scripts/assets/348150/d422a0e7-85aa-4cd9-8822-82fe378ef070)